lib/utils/simnet-connext-channels.ts

```diff
@@ -0,0 +1,196 @@
+import { from, of, empty, Observable, combineLatest, throwError } from 'rxjs';
+import { mergeMap, concat, map, catchError, delay, repeat, retryWhen, take } from 'rxjs/operators';
+import { XudClient } from '../proto/xudrpc_grpc_pb';
+import grpc from 'grpc';
+import http from 'http';
+import { loadXudClient } from '../cli/command';
+import {
+  GetBalanceRequest,
+  GetBalanceResponse,
+  GetInfoRequest,
+  GetInfoResponse,
+  OpenChannelRequest,
+  OpenChannelResponse,
+ } from '../proto/xudrpc_pb';
+
+interface GrpcResponse {
+  toObject: Function;
+}
+
+type Balances = {
+  channelBalance: number;
+  walletBalance: number;
+};
+
+const processResponse = (resolve: Function, reject: Function) => {
+  return (error: grpc.ServiceError | null, response: GrpcResponse) => {
+    if (error) {
+      reject(error);
+    } else {
+      const responseObj = response.toObject();
+      resolve(responseObj);
+    }
+  };
+};
+
+const getBalance = async (client: XudClient, currency?: string): Promise<GetBalanceResponse.AsObject> => {
+  const request = new GetBalanceRequest();
+  if (currency) {
+    request.setCurrency(currency.toUpperCase());
+  }
+  const balances = await new Promise((resolve, reject) => {
+    client.getBalance(request, processResponse(resolve, reject));
+  });
+  return balances as GetBalanceResponse.AsObject;
+};
+
+const openConnextChannel = async (client: XudClient, currency: string, amount: number): Promise<OpenChannelResponse.AsObject> => {
+  const request = new OpenChannelRequest();
+  request.setCurrency(currency.toUpperCase());
+  request.setNodeIdentifier('Connext'); // TODO: to be changed when xud/connext PR is updated to exclude nodeIdentifier
+  request.setAmount(amount);
+  const openChannelResponse = await new Promise((resolve, reject) => {
+    client.openChannel(request, processResponse(resolve, reject));
+  });
+  return openChannelResponse as OpenChannelResponse.AsObject;
+};
+
+const checkBalanceObservable = (client: XudClient, currency: string, minimumWalletBalance: number): Observable<Balances> => {
+  return from(
+    getBalance(client, currency),
+  ).pipe(
+    map((balanceResponse: GetBalanceResponse.AsObject) => {
+      const balanceMap = balanceResponse.balancesMap[0];
+      return {
+        walletBalance: balanceMap[1].walletBalance,
+        channelBalance: balanceMap[1].channelBalance,
+      };
+    }),
+    mergeMap((balances) => {
+      // if the walletBalance is below the minimum threshold we error in
+      // order to trigger the faucet request
+      if (balances.walletBalance < minimumWalletBalance) {
+        return throwError(`balance under the threshold of ${minimumWalletBalance}`);
+      } else {
+        return of(balances);
+      }
+    }),
+    catchError(() => {
```

Comment:
Will this catch errors that are not related to being below balance? Why not put this faucet request directly in the `if (balances.walletBalance < minimumWalletBalance)` block?

Reply:
Very good point. Thanks. I moved this block directly under the `walletBalance` check.